### PR TITLE
Ligand filter bug

### PR DESCRIPTION
**Pull request by maylinnp**
There was a bug in the filters pertaining to ligands (ligand name, max heavy atoms, smarts substructures and position) where the optional 'ligand_operator' was used incorrectly. For example, if the ligand_operator was "OR" then all filter groups (not just ligand filters) would be joined by the logical operator 'OR' in the database query. In the case of 'AND' the bug would be no different than without the bug. The ligand operator is intended to be used when specifying two or more SMARTS pattern be present in a ligand, or for two or more SMARTS position filters, while all other filter groups* are expected to be joined by 'AND'. (*It should be noted that for the filter "ligand_name" if more than one value is given, these will always be joined by 'OR'.)
In addition to the bug fix, the doc strings were updated to be clearer in how the 'ligand_operator' is expected to work, as well as some minor clean ups in the methods pertaining to constructing the ligand filter queries.

### CODE REVIEW

**Comment by rwxayheee** [comment]
Hi @maylinnp 
Here https://github.com/forlilab/Ringtail/blob/ligand_filter_bug/ringtail/storagemanager.py#L3142-L3146
This suggests that "ligand_name" is a list of names (which is different from the documentation but it's useful), and if "ligand_name" is part of the filtering condition, then the selection string ends with OR here. This might cause unexpected precedence and might be causing the original issue I described here

```
            if keyword == "ligand_name":
                for name in filter:
                    if name == "":
                        continue
                    sql_ligand_string += f" L.LigName LIKE '%{name}%' OR"
            if keyword == "ligand_max_atoms" and filter is not None:
                sql_ligand_string += f" mol_num_hvyatms(ligand_rdmol) <= {filter} AND"
```

**Comment by maylinnp** [comment]
@rwxayheee Great find, I rewrote that method more so the logic should be clearer (and more correct) and removed the need for stripping the end of the query string.

File: ringtail/storagemanager.py [hunk 1]
```diff
@@ -3153,46 +3127,49 @@ def _generate_ligand_filtering_query(self, ligand_filters: dict) -> str:
...
-            if kw == "ligand_substruct":
-                for smarts in fils:
+
+        for keyword, filter in ligand_filters.items():
+            if keyword == "ligand_name":
+                # make each name a partial sql string in list format
+                names = [
+                    f"L.LigName LIKE '%{name}%'" for name in filter if filter is not ""
+                ]
+                query_list.append(" OR ".join(names))
```
Comment (rwxayheee) [inline ringtail/storagemanager.py]: If there are multiple name filters, all of them need to be embedded in Parentheses. 
```py
                query_list.append("( "+ " OR ".join(names) + ") ")
```

File: ringtail/storagemanager.py [hunk 2]
```diff
@@ -2608,10 +2608,11 @@ def _generate_result_filtering_query(self, filters_dict):
                             self._ligand_substructure_position_filter(temp_lig_filter)
                         )
                     join_stmnt = " " + lig_filters["ligand_operator"] + " "
```
Comment (rwxayheee) [inline ringtail/storagemanager.py:2614]: I think Parentheses are also needed somewhere after this, to make sure multiple conditions are evaluated by the user specified logical_operator within their group, but eventually the groups (name, substructure, etc.) are always combined by "AND". When we finish the joining of multiple statements within a specific filter group, we could put Parentheses just to be safe.

**Comment by rwxayheee** [comment]
Hi @maylinnp 
Sorry for the late response. I left two comments in the codes and I suggest printing the selection string and checking the order of evaluation to see if it's what's expected. But to be safe, maybe we can just add the parentheses every time we close a specific group of filters (name, struct, etc). 

I'm not very familiar with the other parts that were edited in this PR. But again no pressure on continue working on this, especially if you already have another branch that is more advanced and have alternates to the chemicalite-dependent contents. You could leave this open, or close this as you want, and I can help review the final version with the chemicalite-changes sometime in the near future. Again, no rush at all

**Comment by maylinnp** [comment]
Good catch, I had forgotten one set of parentheses in the ligand filtering method. The other please you mention already has parentheses:
`" AND (" + join_stmnt.join(ligand_substruct_queries) + ")"`

As mentioned yesterday, this is a bug and I plan to release this bug fix shortly. The other work I told you about is not a bug fix and should not be mixed in with this bug fix. Sorry for not making a clearer distinction before.

**Comment by rwxayheee** [comment]
Hi @maylinnp 

There might be some other problem with ligand_substruct_pos filter that made me think the pos queries were wrong. This is how I found out about it: 

```py
rtc.filter(ligand_substruct=["[Oh]C", "C=O"], ligand_name=["C", "3"],ligand_operator="AND", ligand_max_atoms=40)
```

gives me 

```
Number of ligands passing filters: 13
```

While this

```py
rtc.filter(ligand_substruct=["[Oh]C", "C=O"], ligand_name=["C", "3"],ligand_operator="AND", ligand_max_atoms=40, 
           ligand_substruct_pos=[["C=O", 0, 10, 89.474,  84.341,  -7.311]])
``

gives

```
Number of ligands passing filters: 273
```

To do a self-check, you can play with some random filters on any database. Other members in this lab can also help you to check and approve this PR, but if no one does, I can look at it on Tuesday. 

I think a lot of things related to the mol* queries here in the chemicalite-independent version, so I understand if part of this PR isn't really necessary.

**Comment by maylinnp** [comment]
Are you on the ligand_filter_bug branch? -May-Linn22. nov. 2024 kl. 12:25 skrev Amy He ***@***.***>:﻿
Hi @maylinnp
There might be some other problem with ligand_substruct_pos filter that made me think the pos queries were wrong. This is how I found out about it:
rtc.filter(ligand_substruct=["[Oh]C", "C=O"], ligand_name=["C", "3"],ligand_operator="AND", ligand_max_atoms=40)
gives me
Number of ligands passing filters: 13

While this
rtc.filter(ligand_substruct=["[Oh]C", "C=O"], ligand_name=["C", "3"],ligand_operator="AND", ligand_max_atoms=40, 
           ligand_substruct_pos=[["C=O", 0, 10, 89.474,  84.341,  -7.311]])
``
gives
Number of ligands passing filters: 273

To do a self-check, you can play with some random filters on any database. Other members in this lab can also help you to check and approve this PR, but if no one does, I can look at it on Tuesday. 

I think a lot of things related to the mol* queries here in the chemicalite-independent version, so I understand if part of this PR isn't really necessary.

**Comment by rwxayheee** [comment]
Yes, but if no one else can reproduce this maybe I did something wrong. I can upload the input and output

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by rwxayheee** [comment]
[ringtail_ligand_filter.ipynb.txt](https://github.com/user-attachments/files/17876632/ringtail_ligand_filter.ipynb.txt)
[output.db.txt](https://github.com/user-attachments/files/17876633/output.db.txt)
Here you can find the database and the notebook. No rush on this, I can look at it on Tuesday, maybe I did something wrong

**Comment by maylinnp** [comment]
I juts ran you filter (with different x y z coordinates appropriate to by db) on a 2 mill ligand database, and for the least restrictive query I got 100,512 passing ligands:
`rtc.filter(
    ligand_name=["1", "A"],
    ligand_substruct=["C=O", "[C][Oh]"],
    ligand_operator="AND",
    ligand_max_atoms=40,
)`.

For the more restrictive query I got 37,115 passing ligands:
`rtc.filter(
    ligand_name=["1", "A"],
    ligand_substruct=["C=O", "[C][Oh]"],
    ligand_substruct_pos=[
        ["C=O", 1, 5, -27, 21, 30],
    ],
    ligand_operator="AND",
    ligand_max_atoms=40,]
)`

(The more restrictive query also has the same number of passing ligands as if I used the `ligand_operator = "OR"` which makes sense as the main qualification is the smarts pattern in the specified location. The less restrictive filter with ligand operator OR returned over a million ligands.)

I wonder if maybe you didn't pull the latest version of the `ligand_filter_bug` branch?

**Comment by rwxayheee** [comment]
Hi @maylinnp 
I'm at the latest (68b45aa). Could you please take a look at my Jupyter notebook file (input + output)? You can read the database. 
[output.db.txt](https://github.com/user-attachments/files/17878137/output.db.txt)
[ringtail_ligand_filter.ipynb.txt](https://github.com/user-attachments/files/17878140/ringtail_ligand_filter.ipynb.txt)
At the beginning, I was preparing this as part of a workshop tutorial, but we submitted now and it's a not urgent question. So it's really not urgent, and I feel bad for wasting your time on this. Maybe I did something wrong, but it's very confusing that stricter filter returned more passing ligands.

**Comment by maylinnp** [comment]
@rwxayheee found the bug! It had to do with the ligandname filter, I know I did not test that particular filter a lot (clearly not enough). 
Thank you for the thorough testing, sorry it took me a day to squash this (hopefully) last ligand filter bug!
commit ed316be

**Review by rwxayheee — APPROVED** [review]
Thanks again for the nice work. I approve the PR as I can't reproduce the original issue. As my final point, combination of strings isn't the only way to build queries. There might be methods like to establish the query structure and then execute with parameters to make maintenance easier and potentially faster execution, which might align with your plans moving forward.

**Comment by maylinnp** [comment]
@rwxayheee Thank you, glad to hear it! And yes I think you bring up a good point about how to build queries, I will keep your suggestion in mind for next bigger release.

**Pull request merged**